src/scenic/core/object_types.py

```diff
@@ -3,666 +3,1760 @@
+            dynTypes[prop] = ty
+        cls._dynamicProperties = dynTypes
+
+    @classmethod
+    def _withProperties(cls, props, constProps=frozenset()):
+        assert all(reqProp in props for reqProp in cls._defaults)
+        assert all(not needsLazyEvaluation(val) for val in props.values())
+        return cls(_internal=True, _constProps=constProps, **props)
+
+    @classmethod
+    def _withPropertiesDefaults(cls, props, constProps=frozenset()):
+        assert all(not needsLazyEvaluation(val) for val in props.values())
+        return cls(_register=False,_constProps=constProps, **props)
+
+    def __init__(self, *args, _internal=False, _constProps=frozenset(), _register=True,
+                 **kwargs):
+        if _internal:   # Object is being constructed internally; use fast path
+            assert not args
+            for prop, value in kwargs.items():
+                assert not needsLazyEvaluation(value), (prop, value)
+                object.__setattr__(self, prop, value)
+            super().__init__(kwargs.values())
+            self.properties = tuple(sorted(kwargs.keys()))
+            self._propertiesSet = set(self.properties)
+            self._constProps = _constProps
+            return
+
+        # Resolve and apply specifiers
+        specifiers = list(args)
+        for prop, val in kwargs.items():    # kwargs supported for internal use
+            specifiers.append(Specifier("Internal(Kwargs)", {prop: 1}, {prop: val}))
+
+        # Apply specifiers
+        self._applySpecifiers(specifiers)
+
+        # Set up dependencies
+        deps = []
+        for prop in self.properties:
+            assert hasattr(self, prop)
+            val = getattr(self, prop)
+            deps.append(val)
+        super().__init__(deps)
+
+        # Possibly register this object
+        if _register:
+            self._register()
+
+    def _applySpecifiers(self, specifiers, defs=None, overriding=False):
+        # Declare properties dictionary which maps properties to the specifier
+        # that will specify that property.
+        properties = dict()
+
+        # Declare modifying dictionary, which maps properties to a specifier
+        # that will modify that property.
+        modifying = dict()
+
+        # Dictionary mapping properties set so far to the priority with which they have
+        # been set.
+        priorities = dict()
+
+        # Extract default property values dictionary and set of final properties,
+        # unless defs is overriden.
+        if defs is None:
+            defs = self.__class__._defaults
+
+        finals = self.__class__._finalProperties
+
+        # Check for incompatible specifier combinations
+        specifier_names = [spec.name for spec in specifiers]
+
+        if "On" in specifier_names:
+            if collections.Counter(specifier_names)["On"] > 1:
+                raise RuntimeParseError(f'Cannot use "On" specifier to modify "On" specifier')
```

Comment:
This check seems awfully specific. Can we just ban repeated use of *any* specifier?

Reply:
Hmm we could, but this case provides a nice error message and I don't think it would add anything since all the other cases would be caught be the priority checks.

Reply:
I think you can still have a nice error message, e.g. `f'cannot use "{spec}" specifier to modify itself'`. My concern was that the specifier resolution procedure should not need to know about "on" specifically. It's true there aren't any other cases this would catch right now, but there might be more in the future.

Reply:
Good point, I'll add it.